### ISSUE TITLE
unify different vcs prompts

### ISSUE THREAD

**Issue by rogerfc** [fish-shell/fish-shell#4864]
On macOS 10.13.4
Using fish 2.7.1 installed via homebrew, fresh install
Using Terminal, with $TERM=xterm-256color

ISSUE: 
The default VCS prompts functions included (__fish_git_prompt.fish, __fish_hg_prompt.fish, __fish_svn_prompt.fish) produce each a different prompt style. Also, it is not easy to tweak their output to build a prompt of your own. 

For example, using the informative-vcs prompt, I get
![captura de pantalla 2018-03-30 a les 13 16 17](https://user-images.githubusercontent.com/1453692/38136101-8dff5ede-341c-11e8-9c7a-db159dc45088.png)

It would be nice to be able to have:
- Similar default prompts
- Unique set of variables to set prompt contents and colors
- Let the prompt be prined explicitly in the fish_prompt.fish function

**Comment by zanchey** [comment]
One of the original goals of at least the git prompt was compatibility with the bash implementation, which may need to be taken into account.

**Comment by faho** [comment]
This sounds alright, but there's a bunch of minutia to be sorted out.

>Similar default prompts

- Which style should we use? Should it be `|$prompt` or `($prompt)`? The voltage sign or ellipsis?

- Should we distinguish between the different vcsen in some way? `($vcs:$prompt)`? Should we do that always or just when you have more than one installed?

- What about features that aren't in all supported vcsen or that don't quite map? git doesn't have "bookmarks".

>Unique set of variables to set prompt contents and colors

- Which knobs should we offer?

- Should we allow only setting one value for all vcsen or also different values for different ones?

>Let the prompt be prined explicitly in the fish_prompt.fish function

I'm not quite sure what you expect here. Should `fish_git_prompt` only print the `$branch|$state` part, and you add the parentheses later? Or something more complicated?

**Comment by rogerfc** [comment]
I would just use the fish_git_prompt not to print anything but to populate the set of variables. 

This set I see it similar to what the informative_vcs prompt is already using, with __fish_git_prompt_color_dirtystate, __fish_git_prompt_color_branch, ..., and should include:
* vcs type (git,hg,svn), branch, dirty status, etc. to be set by the individual __fish_(git,svn,hg)_prompt function
* default colors, styles, chars, whether some info shown or not, etc. to be set in a single function, e.g. __fish_vcs_prompt

The set shoud be extensive, including all possible features. In case some vcs does not support branches, then the corresponing function leaves the var unset. I can propose a complete set if you wish. 

The prompt function is then responsible to compile all those vars into the printable string. 

If the prompt function or the user does not overwrites any of these vars, I think it's ok to default to the git style for compatibility reasons.

**Comment by rogerfc** [comment]
btw, it is safe to assume there is just zero or one vcs at any given time?
I don't see how one would use many at the same time...

**Comment by faho** [comment]
>is it safe to assume there is just zero or one vcs at any given time?

No. You can use any number of vcsen in the same directory. That by itself won't often occur (though IIUC there are git-hg and similar bridges that allow using one repo with another vcs, so it would show up as both of them).

But what can happen is that you have e.g. a hg repo inside a git repo. A directory is in a git/hg repository if it or any of the parent directories contains a .git/.hg subdirectory, so e.g. for people who store their entire home in git, this would be a regular occurence. And since you can then actually use both commands, we should show information for both.

>I would just use the fish_git_prompt not to print anything but to populate the set of variables.

Interesting idea. So you'd run `fish_git_prompt`, that then sets $branch and $dirty and $untracked and $staged, and then you have something like

```fish
printf '%s|%s%s%s' "$branch" "$dirty" "$staged" "$untracked"
```

in your `fish_prompt` or `fish_right_prompt`?

That would allow us to have a default in `fish_vcs_prompt` which just looks the same as the current function.

---------------

Like @zanchey said, one of the ideas behind `__fish_git_prompt` (which we really should remove the dunderscore from) was that it'd look like a pixel-perfect copy of the official git-prompt. @kballard was rather insistent on that.

Personally, I don't see the value in that - if we can do better than the official one, we should. What I do see the value in is respecting git's prompt _settings_.

But that's something we can discuss.

**Comment by ariasuni** [comment]
This seems linked to #7047.
